{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "17a243a0",
   "metadata": {},
   "source": [
    "## Datasets that are as clean as possible\n",
    "\n",
    "\n",
    "\n",
    "### Test data - Amphimedon and Nematostella  \n",
    "\n",
    "I'm going to use a sponge and a cnido to start with because they are the kinds of organisms I want this program to be useful for (and also they are cool). They also have decent(?) genomes, or at least well-used ones, so maybe they can help me get an idea of how \"clean\" I can get an assembly to be.  \n",
    "\n",
    "I'll downloaded a dataset for each one from the ENA. The Amphimedon is JADHEG01 and the Nematostella is ABAV01.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e59aaab",
   "metadata": {},
   "source": [
    "### Kraken  \n",
    "\n",
    "I've never used kraken (or kraken2 at this point) before, so this is kind of a slow process. I found the documentation (https://github.com/DerrickWood/kraken2/wiki/Manual#special-databases)(which seems to be organized in a non-intuitive way, so that's fun) and I think it has at least most of what I need. Hopefully I can get the rest from the help page and this tutorial I found: https://bioinformaticsworkbook.org/dataAnalysis/Metagenomics/Kraken.html#gsc.tab=0.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e76ce9f9",
   "metadata": {},
   "source": [
    "I looked at a kraken job that Rebecca submitted (presumably sometime in the last six months) and tried to copy some of her settings whenever possible. She used a database that was already downloaded, and I think it is just the standard kraken one, so I'm going to try to use that too. If I get an error that it is not available, I'll figure out what to do next. The settings and command I tried are below.  \n",
    "\n",
    "- CPU time: short  \n",
    "- Memory: 8Gb  \n",
    "- Type of PE: multi-thread  \n",
    "- Number of CPUs: 12  \n",
    "- Job name: amph_kraken\n",
    "\n",
    "`kraken2 --db /data/genomics/db/Kraken/kraken2_db/ --unclassified-out amph_unclassified --report amph_report_kraken --threads $NSLOTS  Amphimedon.fasta`  \n",
    "\n",
    "After I saved it to my computer and transferred it onto Hydra, I changed the module that it connects to. The only one you can access in the QSub Generation Utility is the oldest installation, so I just added the version directory in the module load part manually."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c936b5",
   "metadata": {},
   "source": [
    "Ok, I finally got this running, and updated the kraken script above to reflect the one that functions.  \n",
    "\n",
    "I also separated the std error, because I think it prints a lot of results there (since there does not seem to be a way to specify a main output file), and because it didn't seem to be able to find my input files when I used > to try to capture it in a different file.  \n",
    "\n",
    "After I got results back I ran it on the Nematostella genome also."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a7b5d98",
   "metadata": {},
   "source": [
    "### Kraken output  \n",
    "\n",
    "**Amphimedon**  \n",
    "1422 sequences classified (36.73%)  \n",
    "2449 sequences unclassified (63.27%)  \n",
    "\n",
    "**Nematostella**  \n",
    "15862 sequences classified (26.82%)  \n",
    "43287 sequences unclassified (73.18%)  \n",
    "\n",
    "Since the database contains just viruses, bacteria, and human sequences, it is good that lots of sequences are unclassified. Also, probably lots of the human ones are misclassified, and if you were dealing with a more specific database, many of them would likely hit closer to the mark. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a782e3bd",
   "metadata": {},
   "source": [
    "### Investigating kmers  \n",
    "\n",
    "In the log file (where kraken dumps all the actual results), you can get the specific info for each contig (or sequence, or whatever you put in it). The first letter just stands for classified or unclassified, then they give the name of the sequence/contig, then what category it was assigned to (unclassified, human, bacteria species, etc.) with its taxon id, the length of the sequence/contig, and then how it got divided up in terms of kmers.  \n",
    "\n",
    "For example, `C\tENA|ABAV01000001|ABAV01000001.1\tHomo sapiens (taxid 9606)\t20022\t0:3438 9606:8 0:16542` this is saying that there were 3438 kmers that did not have a match in the database, 8 kmers that matched human (9606), and then 16542 kmers that didn't have a match again. Somehow, this is still being classified as a human, which is pretty wild. Just anecdotally, looking through this, lots of the ones classified as humans look like this. I don't really understand if there is a threshold or something, because others remain unclassified even when kraken finds some kmers that match human sequences. Like this one: `U\tENA|ABAV01000008|ABAV01000008.1\tunclassified (taxid 0)\t4802\t0:508 9606:5 0:4255`. It has five kmers that match a human, but it is unclassified while the above one is not, even though it's really a higher percentage of the whole sequence.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9f0c139",
   "metadata": {},
   "source": [
    "I decided to sequester all of the contigs that were classified as bacterial and viral in their own file.  \n",
    "\n",
    "`grep -v \"unclassified\\|Homo sapiens\" nema_kraken.log > nema_contaminants.tsv`  \n",
    "\n",
    "So I can look at these in isolation if I decide.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a4b7b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "754e1c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "nema = pd.read_csv(\"nema_kraken.log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "416c3ad6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000001_ABAV01000001.1</td>\n",
       "      <td>Homo sapiens (taxid 9606)</td>\n",
       "      <td>20022.0</td>\n",
       "      <td>0:3438 9606:8 0:16542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000002_ABAV01000002.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>10536.0</td>\n",
       "      <td>0:10502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000003_ABAV01000003.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>904.0</td>\n",
       "      <td>0:870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000004_ABAV01000004.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>4548.0</td>\n",
       "      <td>0:4514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000005_ABAV01000005.1</td>\n",
       "      <td>Homo sapiens (taxid 9606)</td>\n",
       "      <td>49951.0</td>\n",
       "      <td>0:11978 9606:2 0:1452 9606:1 0:902 9606:5 0:84...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  class_unclass                      contig_name                      taxon  \\\n",
       "0             C  ENA_ABAV01000001_ABAV01000001.1  Homo sapiens (taxid 9606)   \n",
       "1             U  ENA_ABAV01000002_ABAV01000002.1     unclassified (taxid 0)   \n",
       "2             U  ENA_ABAV01000003_ABAV01000003.1     unclassified (taxid 0)   \n",
       "3             U  ENA_ABAV01000004_ABAV01000004.1     unclassified (taxid 0)   \n",
       "4             C  ENA_ABAV01000005_ABAV01000005.1  Homo sapiens (taxid 9606)   \n",
       "\n",
       "    length                                              kmers  \n",
       "0  20022.0                              0:3438 9606:8 0:16542  \n",
       "1  10536.0                                            0:10502  \n",
       "2    904.0                                              0:870  \n",
       "3   4548.0                                             0:4514  \n",
       "4  49951.0  0:11978 9606:2 0:1452 9606:1 0:902 9606:5 0:84...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nema.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ce2f260b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000003_ABAV01000003.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>904.0</td>\n",
       "      <td>0:870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000004_ABAV01000004.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>4548.0</td>\n",
       "      <td>0:4514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000007_ABAV01000007.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>2191.0</td>\n",
       "      <td>0:2157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000008_ABAV01000008.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>4802.0</td>\n",
       "      <td>0:508 9606:5 0:4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01000010_ABAV01000010.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>972.0</td>\n",
       "      <td>0:938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  class_unclass                      contig_name                   taxon  \\\n",
       "2             U  ENA_ABAV01000003_ABAV01000003.1  unclassified (taxid 0)   \n",
       "3             U  ENA_ABAV01000004_ABAV01000004.1  unclassified (taxid 0)   \n",
       "6             U  ENA_ABAV01000007_ABAV01000007.1  unclassified (taxid 0)   \n",
       "7             U  ENA_ABAV01000008_ABAV01000008.1  unclassified (taxid 0)   \n",
       "9             U  ENA_ABAV01000010_ABAV01000010.1  unclassified (taxid 0)   \n",
       "\n",
       "   length                kmers  \n",
       "2   904.0                0:870  \n",
       "3  4548.0               0:4514  \n",
       "6  2191.0               0:2157  \n",
       "7  4802.0  0:508 9606:5 0:4255  \n",
       "9   972.0                0:938  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nema[nema.length <= 5000].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1713f649",
   "metadata": {},
   "source": [
    "This is showing me the contigs that are below a certain length, but I'm going to try sorting them, so I can see all the ones that are small.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "db1af4e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9471</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009472_ABAV01009472.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9469</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009470_ABAV01009470.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9468</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009469_ABAV01009469.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9472</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009473_ABAV01009473.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9475</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009476_ABAV01009476.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9478</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009479_ABAV01009479.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9466</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009467_ABAV01009467.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9465</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009466_ABAV01009466.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9470</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009471_ABAV01009471.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9463</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009464_ABAV01009464.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9464</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009465_ABAV01009465.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9477</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009478_ABAV01009478.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9467</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009468_ABAV01009468.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9479</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009480_ABAV01009480.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>35.0</td>\n",
       "      <td>A:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9474</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009475_ABAV01009475.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>38.0</td>\n",
       "      <td>A:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9473</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009474_ABAV01009474.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>40.0</td>\n",
       "      <td>A:6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9476</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009477_ABAV01009477.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>49.0</td>\n",
       "      <td>A:15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9462</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01009463_ABAV01009463.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>54.0</td>\n",
       "      <td>A:20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14272</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01014273_ABAV01014273.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>61.0</td>\n",
       "      <td>A:27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19684</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01019685_ABAV01019685.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>119.0</td>\n",
       "      <td>0:85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      class_unclass                      contig_name                   taxon  \\\n",
       "9471              U  ENA_ABAV01009472_ABAV01009472.1  unclassified (taxid 0)   \n",
       "9469              U  ENA_ABAV01009470_ABAV01009470.1  unclassified (taxid 0)   \n",
       "9468              U  ENA_ABAV01009469_ABAV01009469.1  unclassified (taxid 0)   \n",
       "9472              U  ENA_ABAV01009473_ABAV01009473.1  unclassified (taxid 0)   \n",
       "9475              U  ENA_ABAV01009476_ABAV01009476.1  unclassified (taxid 0)   \n",
       "9478              U  ENA_ABAV01009479_ABAV01009479.1  unclassified (taxid 0)   \n",
       "9466              U  ENA_ABAV01009467_ABAV01009467.1  unclassified (taxid 0)   \n",
       "9465              U  ENA_ABAV01009466_ABAV01009466.1  unclassified (taxid 0)   \n",
       "9470              U  ENA_ABAV01009471_ABAV01009471.1  unclassified (taxid 0)   \n",
       "9463              U  ENA_ABAV01009464_ABAV01009464.1  unclassified (taxid 0)   \n",
       "9464              U  ENA_ABAV01009465_ABAV01009465.1  unclassified (taxid 0)   \n",
       "9477              U  ENA_ABAV01009478_ABAV01009478.1  unclassified (taxid 0)   \n",
       "9467              U  ENA_ABAV01009468_ABAV01009468.1  unclassified (taxid 0)   \n",
       "9479              U  ENA_ABAV01009480_ABAV01009480.1  unclassified (taxid 0)   \n",
       "9474              U  ENA_ABAV01009475_ABAV01009475.1  unclassified (taxid 0)   \n",
       "9473              U  ENA_ABAV01009474_ABAV01009474.1  unclassified (taxid 0)   \n",
       "9476              U  ENA_ABAV01009477_ABAV01009477.1  unclassified (taxid 0)   \n",
       "9462              U  ENA_ABAV01009463_ABAV01009463.1  unclassified (taxid 0)   \n",
       "14272             U  ENA_ABAV01014273_ABAV01014273.1  unclassified (taxid 0)   \n",
       "19684             U  ENA_ABAV01019685_ABAV01019685.1  unclassified (taxid 0)   \n",
       "\n",
       "       length  kmers  \n",
       "9471     12.0    0:0  \n",
       "9469     18.0    0:0  \n",
       "9468     21.0    0:0  \n",
       "9472     28.0    0:0  \n",
       "9475     29.0    0:0  \n",
       "9478     30.0    0:0  \n",
       "9466     30.0    0:0  \n",
       "9465     30.0    0:0  \n",
       "9470     31.0    0:0  \n",
       "9463     32.0    0:0  \n",
       "9464     32.0    0:0  \n",
       "9477     33.0    0:0  \n",
       "9467     34.0    0:0  \n",
       "9479     35.0   A:1   \n",
       "9474     38.0   A:4   \n",
       "9473     40.0   A:6   \n",
       "9476     49.0  A:15   \n",
       "9462     54.0  A:20   \n",
       "14272    61.0  A:27   \n",
       "19684   119.0   0:85  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_nema = nema.sort_values(by = [\"length\"])\n",
    "sorted_nema.head(n = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0f5e6bd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48053</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01048054_ABAV01048054.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0:370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55835</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01055836_ABAV01055836.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0:370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41353</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01041354_ABAV01041354.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0:370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18732</th>\n",
       "      <td>U</td>\n",
       "      <td>ENA_ABAV01018733_ABAV01018733.1</td>\n",
       "      <td>unclassified (taxid 0)</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0:370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52029</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01052030_ABAV01052030.1</td>\n",
       "      <td>Homo sapiens (taxid 9606)</td>\n",
       "      <td>405.0</td>\n",
       "      <td>0:313 9606:4 0:1 9606:5 0:48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      class_unclass                      contig_name  \\\n",
       "48053             U  ENA_ABAV01048054_ABAV01048054.1   \n",
       "55835             U  ENA_ABAV01055836_ABAV01055836.1   \n",
       "41353             U  ENA_ABAV01041354_ABAV01041354.1   \n",
       "18732             U  ENA_ABAV01018733_ABAV01018733.1   \n",
       "52029             C  ENA_ABAV01052030_ABAV01052030.1   \n",
       "\n",
       "                           taxon  length                         kmers  \n",
       "48053     unclassified (taxid 0)   404.0                         0:370  \n",
       "55835     unclassified (taxid 0)   404.0                         0:370  \n",
       "41353     unclassified (taxid 0)   404.0                         0:370  \n",
       "18732     unclassified (taxid 0)   404.0                         0:370  \n",
       "52029  Homo sapiens (taxid 9606)   405.0  0:313 9606:4 0:1 9606:5 0:48  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_nema.iloc[200:205]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc2a7074",
   "metadata": {},
   "source": [
    "The smallest contigs look like they are either unclassified or barely getting classified as human. I want to check out the bacteria and virus contigs specifically. I used the same grep line as above that I previously used on Hydra, but on the slightly modified csv I made of the nematostella log file above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "04134a6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000148_ABAV01000148.1</td>\n",
       "      <td>Pandoraea oxalativorans (taxid 573737)</td>\n",
       "      <td>12303.0</td>\n",
       "      <td>0:290 573737:1 0:994 9606:4 0:112 9606:4 13156...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000310_ABAV01000310.1</td>\n",
       "      <td>cellular organisms (taxid 131567)</td>\n",
       "      <td>10228.0</td>\n",
       "      <td>A:1 0:1669 9606:2 0:401 386415:2 0:8119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000479_ABAV01000479.1</td>\n",
       "      <td>Streptomyces sp. TLI_053 (taxid 1855352)</td>\n",
       "      <td>23069.0</td>\n",
       "      <td>0:721 1855352:5 0:155 1855352:5 0:27 1855352:5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000507_ABAV01000507.1</td>\n",
       "      <td>Streptococcus dysgalactiae subsp. equisimilis ...</td>\n",
       "      <td>3767.0</td>\n",
       "      <td>0:773 119602:5 0:1609 9606:4 0:1342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01000535_ABAV01000535.1</td>\n",
       "      <td>Rhodobacteraceae bacterium (taxid 1904441)</td>\n",
       "      <td>22606.0</td>\n",
       "      <td>0:5819 1261031:2 0:10583 1904441:5 0:6163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  class_unclass                      contig_name  \\\n",
       "0             C  ENA_ABAV01000148_ABAV01000148.1   \n",
       "1             C  ENA_ABAV01000310_ABAV01000310.1   \n",
       "2             C  ENA_ABAV01000479_ABAV01000479.1   \n",
       "3             C  ENA_ABAV01000507_ABAV01000507.1   \n",
       "4             C  ENA_ABAV01000535_ABAV01000535.1   \n",
       "\n",
       "                                               taxon   length  \\\n",
       "0             Pandoraea oxalativorans (taxid 573737)  12303.0   \n",
       "1                  cellular organisms (taxid 131567)  10228.0   \n",
       "2           Streptomyces sp. TLI_053 (taxid 1855352)  23069.0   \n",
       "3  Streptococcus dysgalactiae subsp. equisimilis ...   3767.0   \n",
       "4         Rhodobacteraceae bacterium (taxid 1904441)  22606.0   \n",
       "\n",
       "                                               kmers  \n",
       "0  0:290 573737:1 0:994 9606:4 0:112 9606:4 13156...  \n",
       "1            A:1 0:1669 9606:2 0:401 386415:2 0:8119  \n",
       "2  0:721 1855352:5 0:155 1855352:5 0:27 1855352:5...  \n",
       "3                0:773 119602:5 0:1609 9606:4 0:1342  \n",
       "4          0:5819 1261031:2 0:10583 1904441:5 0:6163  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "micros = pd.read_csv(\"nema_micros.csv\")\n",
    "micros.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "33af5389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01015544_ABAV01015544.1</td>\n",
       "      <td>other sequences (taxid 28384)</td>\n",
       "      <td>280.0</td>\n",
       "      <td>28384:25 1:158 0:3 1:5 0:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1442</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01048353_ABAV01048353.1</td>\n",
       "      <td>Pseudomonas sp. CC6-YY-74 (taxid 1930532)</td>\n",
       "      <td>444.0</td>\n",
       "      <td>0:66 135621:12 1534110:5 286:3 1534110:3 0:62 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01023546_ABAV01023546.1</td>\n",
       "      <td>Clostridium botulinum (taxid 1491)</td>\n",
       "      <td>461.0</td>\n",
       "      <td>0:22 526969:5 131567:49 0:10 131567:1 0:29 131...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01043368_ABAV01043368.1</td>\n",
       "      <td>Clostridium baratii str. Sullivan (taxid 1415775)</td>\n",
       "      <td>521.0</td>\n",
       "      <td>0:103 262:1 0:30 1415775:2 0:351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1866</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01056593_ABAV01056593.1</td>\n",
       "      <td>Actinoalloteichus sp. AHMU CJ021 (taxid 2072503)</td>\n",
       "      <td>537.0</td>\n",
       "      <td>0:342 2072503:5 2:5 2072503:2 2:5 2072503:2 2:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01045675_ABAV01045675.1</td>\n",
       "      <td>Tenacibaculum todarodis (taxid 1850252)</td>\n",
       "      <td>546.0</td>\n",
       "      <td>0:191 1850252:4 0:8 1312072:2 0:14 1850252:15 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01051907_ABAV01051907.1</td>\n",
       "      <td>Clostridium botulinum (taxid 1491)</td>\n",
       "      <td>550.0</td>\n",
       "      <td>0:11 1491:5 0:64 9606:4 0:69 1491:5 0:3 1491:2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01005445_ABAV01005445.1</td>\n",
       "      <td>other sequences (taxid 28384)</td>\n",
       "      <td>574.0</td>\n",
       "      <td>28384:29 1:3 0:508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1862</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01056569_ABAV01056569.1</td>\n",
       "      <td>Clostridium botulinum (taxid 1491)</td>\n",
       "      <td>589.0</td>\n",
       "      <td>0:303 1491:4 131567:3 0:5 1491:3 0:35 2057025:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1863</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01056570_ABAV01056570.1</td>\n",
       "      <td>Bacillus cereus m1293 (taxid 526973)</td>\n",
       "      <td>599.0</td>\n",
       "      <td>0:53 A:34 9606:2 0:45 9606:7 0:4 9606:17 0:22 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1774</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01055177_ABAV01055177.1</td>\n",
       "      <td>Bacillus cereus m1293 (taxid 526973)</td>\n",
       "      <td>619.0</td>\n",
       "      <td>A:32 0:4 526973:32 A:34 0:149 526973:5 0:2 960...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01055659_ABAV01055659.1</td>\n",
       "      <td>Bacillus cereus m1293 (taxid 526973)</td>\n",
       "      <td>624.0</td>\n",
       "      <td>0:55 526973:10 131567:6 0:3 526973:3 0:28 5269...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1269</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01043758_ABAV01043758.1</td>\n",
       "      <td>Clostridium botulinum (taxid 1491)</td>\n",
       "      <td>625.0</td>\n",
       "      <td>0:173 526973:4 0:5 1239:4 0:2 131567:5 0:127 5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1499</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01049521_ABAV01049521.1</td>\n",
       "      <td>Microbulbifer agarilyticus (taxid 260552)</td>\n",
       "      <td>625.0</td>\n",
       "      <td>0:74 351679:1 0:5 1236:2 400668:5 2:4 400668:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01013496_ABAV01013496.1</td>\n",
       "      <td>Sphaerochaeta globosa str. Buddy (taxid 158189)</td>\n",
       "      <td>638.0</td>\n",
       "      <td>0:304 158189:1 0:263 158189:1 0:35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01050015_ABAV01050015.1</td>\n",
       "      <td>Bacillus cereus m1293 (taxid 526973)</td>\n",
       "      <td>642.0</td>\n",
       "      <td>0:10 131567:5 0:1 131567:8 0:11 131567:1 9606:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1747</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01054767_ABAV01054767.1</td>\n",
       "      <td>Escherichia coli SE11 (taxid 409438)</td>\n",
       "      <td>645.0</td>\n",
       "      <td>0:343 9606:1 0:13 409438:5 0:37 409438:5 0:37 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1260</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01043594_ABAV01043594.1</td>\n",
       "      <td>Bacteroidetes (taxid 976)</td>\n",
       "      <td>647.0</td>\n",
       "      <td>0:350 28453:5 0:88 308865:5 0:153 1236:3 0:9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1773</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01055176_ABAV01055176.1</td>\n",
       "      <td>Bacillus cereus m1293 (taxid 526973)</td>\n",
       "      <td>657.0</td>\n",
       "      <td>131567:48 0:10 131567:1 0:29 131567:1 0:11 131...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1259</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01043593_ABAV01043593.1</td>\n",
       "      <td>Nonlabens sp. MB-3u-79 (taxid 2058134)</td>\n",
       "      <td>659.0</td>\n",
       "      <td>0:150 2058134:5 49546:6 2058134:2 0:208 96345:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     class_unclass                      contig_name  \\\n",
       "355              C  ENA_ABAV01015544_ABAV01015544.1   \n",
       "1442             C  ENA_ABAV01048353_ABAV01048353.1   \n",
       "632              C  ENA_ABAV01023546_ABAV01023546.1   \n",
       "1250             C  ENA_ABAV01043368_ABAV01043368.1   \n",
       "1866             C  ENA_ABAV01056593_ABAV01056593.1   \n",
       "1329             C  ENA_ABAV01045675_ABAV01045675.1   \n",
       "1632             C  ENA_ABAV01051907_ABAV01051907.1   \n",
       "130              C  ENA_ABAV01005445_ABAV01005445.1   \n",
       "1862             C  ENA_ABAV01056569_ABAV01056569.1   \n",
       "1863             C  ENA_ABAV01056570_ABAV01056570.1   \n",
       "1774             C  ENA_ABAV01055177_ABAV01055177.1   \n",
       "1796             C  ENA_ABAV01055659_ABAV01055659.1   \n",
       "1269             C  ENA_ABAV01043758_ABAV01043758.1   \n",
       "1499             C  ENA_ABAV01049521_ABAV01049521.1   \n",
       "305              C  ENA_ABAV01013496_ABAV01013496.1   \n",
       "1525             C  ENA_ABAV01050015_ABAV01050015.1   \n",
       "1747             C  ENA_ABAV01054767_ABAV01054767.1   \n",
       "1260             C  ENA_ABAV01043594_ABAV01043594.1   \n",
       "1773             C  ENA_ABAV01055176_ABAV01055176.1   \n",
       "1259             C  ENA_ABAV01043593_ABAV01043593.1   \n",
       "\n",
       "                                                  taxon  length  \\\n",
       "355                       other sequences (taxid 28384)   280.0   \n",
       "1442          Pseudomonas sp. CC6-YY-74 (taxid 1930532)   444.0   \n",
       "632                  Clostridium botulinum (taxid 1491)   461.0   \n",
       "1250  Clostridium baratii str. Sullivan (taxid 1415775)   521.0   \n",
       "1866   Actinoalloteichus sp. AHMU CJ021 (taxid 2072503)   537.0   \n",
       "1329            Tenacibaculum todarodis (taxid 1850252)   546.0   \n",
       "1632                 Clostridium botulinum (taxid 1491)   550.0   \n",
       "130                       other sequences (taxid 28384)   574.0   \n",
       "1862                 Clostridium botulinum (taxid 1491)   589.0   \n",
       "1863               Bacillus cereus m1293 (taxid 526973)   599.0   \n",
       "1774               Bacillus cereus m1293 (taxid 526973)   619.0   \n",
       "1796               Bacillus cereus m1293 (taxid 526973)   624.0   \n",
       "1269                 Clostridium botulinum (taxid 1491)   625.0   \n",
       "1499          Microbulbifer agarilyticus (taxid 260552)   625.0   \n",
       "305     Sphaerochaeta globosa str. Buddy (taxid 158189)   638.0   \n",
       "1525               Bacillus cereus m1293 (taxid 526973)   642.0   \n",
       "1747               Escherichia coli SE11 (taxid 409438)   645.0   \n",
       "1260                          Bacteroidetes (taxid 976)   647.0   \n",
       "1773               Bacillus cereus m1293 (taxid 526973)   657.0   \n",
       "1259             Nonlabens sp. MB-3u-79 (taxid 2058134)   659.0   \n",
       "\n",
       "                                                  kmers  \n",
       "355                         28384:25 1:158 0:3 1:5 0:55  \n",
       "1442  0:66 135621:12 1534110:5 286:3 1534110:3 0:62 ...  \n",
       "632   0:22 526969:5 131567:49 0:10 131567:1 0:29 131...  \n",
       "1250                   0:103 262:1 0:30 1415775:2 0:351  \n",
       "1866  0:342 2072503:5 2:5 2072503:2 2:5 2072503:2 2:...  \n",
       "1329  0:191 1850252:4 0:8 1312072:2 0:14 1850252:15 ...  \n",
       "1632  0:11 1491:5 0:64 9606:4 0:69 1491:5 0:3 1491:2...  \n",
       "130                                  28384:29 1:3 0:508  \n",
       "1862  0:303 1491:4 131567:3 0:5 1491:3 0:35 2057025:...  \n",
       "1863  0:53 A:34 9606:2 0:45 9606:7 0:4 9606:17 0:22 ...  \n",
       "1774  A:32 0:4 526973:32 A:34 0:149 526973:5 0:2 960...  \n",
       "1796  0:55 526973:10 131567:6 0:3 526973:3 0:28 5269...  \n",
       "1269  0:173 526973:4 0:5 1239:4 0:2 131567:5 0:127 5...  \n",
       "1499  0:74 351679:1 0:5 1236:2 400668:5 2:4 400668:1...  \n",
       "305                  0:304 158189:1 0:263 158189:1 0:35  \n",
       "1525  0:10 131567:5 0:1 131567:8 0:11 131567:1 9606:...  \n",
       "1747  0:343 9606:1 0:13 409438:5 0:37 409438:5 0:37 ...  \n",
       "1260       0:350 28453:5 0:88 308865:5 0:153 1236:3 0:9  \n",
       "1773  131567:48 0:10 131567:1 0:29 131567:1 0:11 131...  \n",
       "1259  0:150 2058134:5 49546:6 2058134:2 0:208 96345:...  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_micros = micros.sort_values(by = [\"length\"])\n",
    "sorted_micros.head(n = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1e95753d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_unclass</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>length</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01002553_ABAV01002553.1</td>\n",
       "      <td>Proteobacteria (taxid 1224)</td>\n",
       "      <td>63096.0</td>\n",
       "      <td>0:17121 1620215:1 0:4106 1002804:1 0:41833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01012069_ABAV01012069.1</td>\n",
       "      <td>Burkholderiales (taxid 80840)</td>\n",
       "      <td>77005.0</td>\n",
       "      <td>0:2551 9606:6 0:4881 9606:4 0:10216 9606:45 0:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01004701_ABAV01004701.1</td>\n",
       "      <td>Edwardsiella (taxid 635)</td>\n",
       "      <td>79998.0</td>\n",
       "      <td>0:1716 9606:2 0:1973 9606:1 0:30 131567:2 0:94...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>C</td>\n",
       "      <td>ENA_ABAV01014257_ABAV01014257.1</td>\n",
       "      <td>Edwardsiella (taxid 635)</td>\n",
       "      <td>108723.0</td>\n",
       "      <td>0:12075 9606:10 0:1626 9606:2 0:2 9606:5 0:206...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2119</th>\n",
       "      <td>= Thu Aug 26 10:44:19 EDT 2021 job nema_kraken...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          class_unclass  \\\n",
       "62                                                    C   \n",
       "264                                                   C   \n",
       "112                                                   C   \n",
       "323                                                   C   \n",
       "2119  = Thu Aug 26 10:44:19 EDT 2021 job nema_kraken...   \n",
       "\n",
       "                          contig_name                          taxon  \\\n",
       "62    ENA_ABAV01002553_ABAV01002553.1    Proteobacteria (taxid 1224)   \n",
       "264   ENA_ABAV01012069_ABAV01012069.1  Burkholderiales (taxid 80840)   \n",
       "112   ENA_ABAV01004701_ABAV01004701.1       Edwardsiella (taxid 635)   \n",
       "323   ENA_ABAV01014257_ABAV01014257.1       Edwardsiella (taxid 635)   \n",
       "2119                              NaN                            NaN   \n",
       "\n",
       "        length                                              kmers  \n",
       "62     63096.0         0:17121 1620215:1 0:4106 1002804:1 0:41833  \n",
       "264    77005.0  0:2551 9606:6 0:4881 9606:4 0:10216 9606:45 0:...  \n",
       "112    79998.0  0:1716 9606:2 0:1973 9606:1 0:30 131567:2 0:94...  \n",
       "323   108723.0  0:12075 9606:10 0:1626 9606:2 0:2 9606:5 0:206...  \n",
       "2119       NaN                                                NaN  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_micros.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4bcf4e2",
   "metadata": {},
   "source": [
    "It looks like kraken is classifying these contigs based on a very small number of kmers in many if not all cases. This might be because it's really designed to work with reads (Illumina ones) not with longer sequences that come from long read tech or assembly. So I think my next move (after the new, correct database is up and running) is to map the reads for these assemblies onto the assembled contigs, and run kraken on all the reads that map. Then I'll be able to compare that output to this (redone with the same database) to see how comparable they are. Could be that a read would get classified really well, but it looks ridiculous in the middle of a hugh contig. Definitely some contaminating reads could have made it into the assembly itself, so it would be good to know how they behave."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5d38208",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
